content/cloud-hosting/azure/azure-service-bus.md

```diff
@@ -0,0 +1,254 @@
+
+![hostname](https://miro.medium.com/max/1400/1*aWa98oXGDpTUfPNjhB3iiw.png)
+
+# การติดตั้งและรัน RabbitMQ ในเครื่อง Local เพื่อทำ Message Queue
+
+1.ดาวโหลด dockerfile และ docker-compose.yaml ตามที่อธิบายตอนต้น มาที่เครื่อง แล้วใช้คำสั่ง docker-compose up -d เพื่อรัน RabbitMQ, ถ้าสำเร็จ ควรจะขึ้นตามกรอบสีแดง
+
+![](https://miro.medium.com/max/1400/1*qCgqRzjNXAyvsRwqFnUgeQ.png)
+
+2.ถ้าเรียบร้อยดี ให้ลองเข้า browser ที่ URL http://localhost:15672/ มันควรจะแสดงหน้าจอ ดังนี้ ขึ้นมา
+
+![](https://miro.medium.com/max/1400/1*SOEsBM-JL2K1vRz3ZXK37Q.png)
+
+3.ใส่ username = admin, password = admin เข้าไป จะพบหน้าจอ overview ดังนี้
+
+![](https://miro.medium.com/max/1400/1*16pW6h1CjtBJhN3VWluZvQ.png)
+
+4.ที่ด้านบนสุด, กดเมนู Queues แล้วทำการกรอกข้อมูล เพื่อทำการสร้าง Queue ชื่อ poc01 ขึ้นมา ดังรูป
+
+![](https://miro.medium.com/max/1400/1*D9yVvQWCEPymjAY3B06x9w.png)
+
+5.เข้าสู่เมนู admin เพื่อตรวจสอบและกำหนดค่า username / password ดังนี้ โดยคุณสามารถ เพิ่มผู้ใช้งานใหม่ในเมนูนี้, อย่างไรก็ตาม สำหรับการทดสอบนี้ เราจะใช้ username =admin, password = admin ในการทดสอบเชื่อมต่อ
+
+![](https://miro.medium.com/max/1400/1*zrxI0xAF-Fz5yUP6MYC9Lw.png)
+
+# การเขียน code เพื่อเชื่อมต่อกับ Azure Service Bus หรือ RabbitMQ
+
+ผมได้ทำการ commit ตัวอย่าง code ไปเก็บไว้ที่
+
+ ```
+https://github.com/nakornttss/azureservicebusrabmitmqamqp
+ ```
+
+ซึ่งคุณสามารถ Clone ลงมาดูได้ โดยในชุดของ source code จะประกอบไปด้วย โปรแกรมดังนี้
+
+Thesender ใช้สำหรับส่ง message ไปยัง Message Queue มีดังนี้
+
+ ```cs 
+using System;
+using Amqp;
+using Amqp.Framing;
+using System.Net;
+var protocal = "amqp";              // use amqps for Azure Service Bus, use amqp for RabmitMQ in local
+var hostname = "localhost:5672";    // use localhost:5672 for RabmitMQ in local
+var username = WebUtility.UrlEncode("");   // Azure Service Bus is PolicyName
+var password = WebUtility.UrlEncode("");   // Azure Service Bus is Primary Key or secondary key
+// Connection string to connect Azure Service Bus or RabmitMQ
+var connectionString = $"{protocal}://{username}:{password}@{hostname}/";
+// Make connection
+var connection = new Connection(new Address(connectionString));
+// Make session
+var amqpSession = new Session(connection);
+// set sender id
+var senderSubscriptionId = "thedemo.amqp.sender";
+// this is queue name
+var queueName = "poc01";
+// Make Link
+var sender = new SenderLink(amqpSession, senderSubscriptionId, queueName);
+for (var i = 0; i < 10; i++)
+{
+// Create new message
+var message = new Message($"Hello Message Queue {i}");
+// Set message id
+message.Properties = new Properties() { MessageId = Guid.NewGuid().ToString() };
+// Add custom property to message
+message.ApplicationProperties = new ApplicationProperties();
+message.ApplicationProperties["Message.Type.Method"] = "Demo Property";
+// Send message to service bus
+sender.Send(message);
+}
+ ```
+
+Thereciever ใช้สำหรับรับ message จาก Message Queue มีดังนี้
+
+ ```cs 
```

Comment:
Please use 4 spaces indent for this C# code block and start the first indent with 4 space.
Here is the example:

````
```cs 
     static void OnMessageCallback(IReceiverLink receiver, Amqp.Message message)
    {
        try
        {
            // Get custom property from message
            var messageType = message.ApplicationProperties["Message.Type.Method"];
            // Read message body
            var rawBody = message.Body;
            Console.WriteLine(rawBody.ToString() + " " + messageType);
            // Accept message once compled reading
            receiver.Accept(message);
        }
        catch (Exception ex)
        {
            // Reject message in case got any error
            receiver.Reject(message);
            Console.WriteLine(ex);
        }
    }
 
```
````